### ISSUE TITLE
A code issue about force plugin

### ISSUE THREAD

**Issue by yongyouhe** [openmm/openmm#4514]
Hi, I am trying to write a force plugin. Simply put, I want to apply a force to each particle, with a force parameter that varies at each step (cannot use customCVForce or customExternalForce). However, when I simulated a simple NaCl system with the Langevin integrator for a test, I found that this force is calculated twice at each step. I am confused about this and would appreciate some guidance.
The test system only has two ions: Na+ and Cl-, OpenMM version is 7.7.0.
In the execute() function of commonCalcForceKernel, the forces are calculated and added to forcebuffers, and the energy is calculated. The following is the sample code of the kernel function. 
```
KERNEL void addForce(GLOBAL mm_long* forcebuffers, int bufferSize, GLOBAL real4* RESTRICT posq, real muz, real fscale) {
    for (int index = GLOBAL_ID; index < NUM_PARTICLECORR; index += GLOBAL_SIZE) {
        real force1 = (real) (posq[index].w*muz*fscale);
        real force_old = (real) (forcebuffers[index+PADDED_NUM_ATOMS*2]/(real)(0x100000000));
        printf("index=%d, force_z=%f, force_old=%f, posq[index]=%f %f %f %f, muz=%f\n", 
                 index, force1, force_old, posq[index].x, posq[index].y, posq[index].z, posq[index].w, muz);
        forcebuffers[index+PADDED_NUM_ATOMS*2] += (mm_long) (posq[index].w*muz*fscale*0x100000000);
        real force_new = (real)(forcebuffers[index+PADDED_NUM_ATOMS*2]/(real)(0x100000000));
        printf("index=%d, force_z=%f, force_new=%f, posq[index]=%f %f %f %f\n", 
                 index, force1, force_new, posq[index].x, posq[index].y, posq[index].z, posq[index].w);
    }
}
```
At one step, the output is as follows. The atom of index=0 is Na+, and the atom of index=1 is Cl-. The energy is calculated and output when includeEnergy is true.
```
Step 1110
index=0, force_z=0.021701, force_old=0.000000, posq[index]=1.786144 1.985374 1.338361 1.000000, muz=-0.002096
index=1, force_z=-0.021701, force_old=0.000000, posq[index]=1.750189 2.170167 1.343860 -1.000000, muz=-0.002096
index=0, force_z=0.021701, force_new=0.021701, posq[index]=1.786144 1.985374 1.338361 1.000000
index=1, force_z=-0.021701, force_new=-0.021701, posq[index]=1.750189 2.170167 1.343860 -1.000000
The energy of added force is 0
Step 1110
index=0, force_z=0.021701, force_old=-13.897074, posq[index]=1.786144 1.985374 1.338361 1.000000, muz=-0.002096
index=1, force_z=-0.021701, force_old=13.905775, posq[index]=1.750189 2.170167 1.343860 -1.000000, muz=-0.002096
index=0, force_z=0.021701, force_new=-13.875373, posq[index]=1.786144 1.985374 1.338361 1.000000
index=1, force_z=-0.021701, force_new=13.884075, posq[index]=1.750189 2.170167 1.343860 -1.000000
The energy of added force is 2.7059e-05
```
I am confused about why the force was calculated twice, and why the forcebuffer is 0 the first time at each step. Was this force_z applied to each particle during the first calculation at each step?

Best regards.

**Comment by peastman** [comment]
I don't see the force being calculated twice.  Each time through the loop it prints two lines (one with `force_old` and one with `force_new`).  And `NUM_PARTICLECORR` is clearly 2, since there are lines with `index=0` and `index=1`.  So there should be four lines printed for each step, which is what it shows.  Am I missing something?

**Comment by yongyouhe** [comment]
> I don't see the force being calculated twice. Each time through the loop it prints two lines (one with `force_old` and one with `force_new`). And `NUM_PARTICLECORR` is clearly 2, since there are lines with `index=0` and `index=1`. So there should be four lines printed for each step, which is what it shows. Am I missing something?

But, there are two identical steps 1110 being output. Is it normal to output twice? Then why are force_old (0.0 and -13.897074 for index=0) and force_new (0.021701 and -13.875373 for index=0) different each time they are output?

**Comment by peastman** [comment]
Can you show your code that executes it?  If the force is being evaluated twice, there's generally a reason for it.

**Comment by yongyouhe** [comment]
> Can you show your code that executes it? If the force is being evaluated twice, there's generally a reason for it.

Sorry to reply so late. I had found the problem -- I made a simple mistake when writing the simulation run script. For testing purposes, I ran the simulation for one step and then output the energy using `getState()` each time. So, it is the function `getState()` that leads to the second output of `Step 1110`. 

Thanks for your reply!
Best regards.
